shotover-proxy/src/transforms/cassandra/sink_cluster/node_pool.rs

```diff
@@ -0,0 +1,86 @@
+use super::routing_key::calculate;
+use super::token_map::TokenMap;
+use crate::transforms::cassandra::sink_cluster::node::CassandraNode;
+use cassandra_protocol::frame::message_execute::BodyReqExecuteOwned;
+use cassandra_protocol::frame::message_result::PreparedMetadata;
+use cassandra_protocol::frame::Version;
+use cassandra_protocol::token::Murmur3Token;
+use cassandra_protocol::types::CBytesShort;
+use rand::prelude::*;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+pub struct NodePool {
+    prepared_metadata: HashMap<CBytesShort, PreparedMetadata>,
+    token_map: TokenMap,
+    pub nodes: Vec<CassandraNode>,
+}
+
+impl NodePool {
+    pub fn new(nodes: Vec<CassandraNode>) -> Self {
+        Self {
+            token_map: TokenMap::new(nodes.as_slice()),
+            nodes,
+            prepared_metadata: HashMap::new(),
+        }
+    }
+
+    pub fn set_nodes(&mut self, nodes: Vec<CassandraNode>) {
+        self.nodes = nodes;
+        self.token_map = TokenMap::new(self.nodes.as_slice());
+    }
+
+    pub fn add_prepared_result(&mut self, id: CBytesShort, metadata: PreparedMetadata) {
+        self.prepared_metadata.insert(id, metadata);
+    }
+
+    pub fn random_node(&mut self, rng: &mut SmallRng) -> &mut CassandraNode {
+        self.nodes
+            .iter_mut()
+            .filter(|x| x.is_up)
+            .choose(rng)
+            .unwrap()
+    }
+
+    pub fn get_random_node_in_dc_rack(
+        &mut self,
+        rack: &String,
+        rng: &mut SmallRng,
+    ) -> &CassandraNode {
+        self.nodes
+            .iter()
+            .filter(|x| x.rack == *rack && x.is_up)
+            .choose(rng)
+            .unwrap()
+    }
+
+    /// Get a token routed replica node for the supplied execute message (if exists)
+    pub fn replica_node(
+        &mut self,
+        execute: &BodyReqExecuteOwned,
+        version: &Version,
+    ) -> Option<&mut CassandraNode> {
+        let metadata = self.prepared_metadata.get(&execute.id).unwrap();
```

Comment:
These unwraps should return errors instead right?

Reply:
Should we error or just return node for the replica?

Reply:
If the client or server is misbehaving then we should return an error.
If its an expected case where we cant find a replica_node then I guess we should return None?

Reply:
Hmm, I could see both occurring. Imagine we have multiple Shotover nodes, the query is prepared on one, but sent to another. The prepared metadata won't be cached.

Reply:
Theres still a few unwraps in this function btw

Reply:
Just leaving a note here: We decided to handle the prepared metadata unwrap by returning an UNPREPARED error to force the driver to give us the metadata.